packages/grpcweb-transport/spec/grpc-web-transport.spec.ts

```diff
@@ -578,7 +610,35 @@ describe('GrpcWebFetchTransport', () => {
             }
         });
 
-        it('throws `DATA_LOSS` if it does not get a trailer', async () => {
+        it('throws `INTERNAL` if empty response body and no status in headers', async () => {
+            const { next, call } = getNextCall();
+            await next([], true);
+            try {
+                await call;
+                fail('should fail');
+            } catch (e) {
+                expect(e).toBeInstanceOf(RpcError);
+                expect(e.code).toBe(GrpcStatusCode[GrpcStatusCode.INTERNAL]);
+            }
+        });
+
+        it('success if empty response body and OK status in headers', async () => {
```

Comment:
According to https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md#responses
>Most responses are expected to have both headers and trailers but Trailers-Only is permitted for calls that produce an immediate error.

Thus I would expect any "Trailers-Only" response to throw an `RpcError`.

Reply:
This is of course logical for non OK status codes. But in practice, it will be strange if received code OK will be accepted as error on client side.
Maybe initially this case (sending from server only status OK with empty body in stream) is strange, because I think normally server should send something like NOT_FOUND code if there is nothing to send in stream.
But at the moment exactly this handling is implemented in the official implementation of the protocol in https://github.com/grpc/grpc-web

Reply:
That sentence in the spec is frustratingly ambiguous. And unfortunately none of the official implementation tests actually assert what should happen when getting an OK trailers-only unary RPC response: https://github.com/grpc/grpc-web/blob/master/packages/grpc-web/test/generated_code_test.js

Though searching through issues on grpc/grpc I get the sense that it is not allowed for the server to send an OK status without a message in a unary RPC. From https://github.com/grpc/grpc-java/issues/2785#issuecomment-286504286:
>Unary must either error or have a message.
>
>The server does not enforce "unary must have one message." If the application closes with OK but no messages, this should become an error of some sort (sent to the client). This would ideally be done in ServerCallImpl.

From https://github.com/grpc/grpc/issues/8010#issuecomment-245759318:
>Is the server calling `StreamObserver.onCompleted()` (or `ServerCall.close(Status.OK,...)`) without first sending a message?
>
>[...] no problem. It made us realize we weren't verifying that on server-side.

From https://github.com/grpc/grpc-java/issues/2243:
>It [appears](https://github.com/grpc/grpc-java/blob/v1.0.0/core/src/main/java/io/grpc/internal/ServerCallImpl.java#L177) that a server can respond OK to a unary RPC without any detection on the server-side. It'd be nice to throw an exception if close(OK) is called without first sending a message for a UNARY response, although that may be against our API stability. At the very least, we should convert the OK into a Status saying the server is broken.

Reply:
For unary RPC, I have not changed the final behavior and in this case there will be DATA_LOSS error code. See test at line 313.

In general, while there is no unambiguous interpretation yet of this case with streams, it is better to have an implementation compatible with other libraries so that it is easier for developers to migrate from one implementation to another.

Reply:
Ahhh, my mistake. This makes sense for `serverStreaming`.